### PR TITLE
Try/group block custom text color

### LINKED ISSUE #18530: Group Block: Add text color to color settings
Hi there!

Let me preface this by saying that I started using the Group block recently and it makes my layouts a lot easier to manage.

**Problem:**
The ability to add a background colour is very useful, but having to target each individual child block of the group to set a text colour is frustrating (and when those blocks don't have that option, it makes things worse as I have to add classes onto them manually).

Moreover, if the group has a dark background color, the placeholder blocks are pretty much invisible, with text and (+) buttons (idle state) being black on dark:

Clip of a Gutenberg Group block with a dark background where a Paragraph block has black text, making it impossible to see what's displayed inside

**Desired solution**
I wish could define a top-level text color alongside the background color to have a cohesive context, and let my styles cascade that text colour into the child blocks.

**Considered alternatives**
For now I'm applying that colour override to all children blocks but it's a tedious task, especially if I have to edit them.

Please let me know if this is a duplicate, or if there is some setting I can leverage to improve my workflow. I am running the latest version of WordPress, 5.3.

Many thanks!

Best
Chris

### LINKED ISSUE #37051: Editor: When picking a custom color dynamically set a text color with good contrast
If I set the background color of a group block to predefined colors, the text color inside of the group block is set to a color with an appropriate contrast (as chosen by the theme css classes `has-secondary-background`)

However, if I set the background color of a group block to a custom color, the text color always switches to dark grey. This means I cannot set a custom dark background color as the text is unreadable with no way to change the color.

Let's see if Gutenberg core would be interested in an enhancement where when picking a custom background color we also dynamically set a text color with good contrast.

### PR DESCRIPTION
**Pull request by glendaviesnz**
## Description

With the current group block, if a custom background colour is selected that does not contrast well with the text in child blocks there is no way to rectify this other than applying a suitable text colour to every child block within the group 

This PR adds a  text colour selector to the Group Block inspector controls in addition to the background colour selector. It also refactors the block to use the new useColors hook instead of withColours

The text colour gets applied to the children by the default css cascade, so applying a text colour to to a child block, at any level of nesting, will override the colour set on the parent group block, and moving the child block out of the parent will also remove the colour. 

Possible additions for follow on PRs, based on feedback to-date:

- Indicate in some way in the child component that a colour is being inherited from the parent
- Automatically select a contrasting text colour if a custom background colour is selected - but it would probably make more sense to add this as an option in the useColors hook than in this specific component

## How has this been tested?

Tested manually by: 

Adding children to a group block at multiple levels of nesting and making sure that text colour apply to group block is applied to text of children, and that any colour applied to child blocks overrides text colour from parent group block
Adding a group block from existing master branch and then opening/editing/saving to make sure there are no backwards compatibility issues.

If there is approval to proceed with this I will see what automated tests might make sense to add.

## Screenshots 

Before:

tcolour-before

After:

manual-text

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards. 
- [  ] My code has proper inline documentation.
- [  ] I've included developer documentation if appropriate. 
- [  ] I've updated all React Native files affected by any refactorings/renamings in this PR.

Fixes https://github.com/Automattic/wp-calypso/issues/37051
Closes #18530

### CODE REVIEW

**Comment by glendaviesnz** [comment]
Adding this as a draft PR at this stage to see if there is interest in adding this to core. If there is interest then I will refine this, fix tests, documentation, etc.

**Comment by glendaviesnz** [comment]
On thinking about this further, it is probably better to just go with the same colour setting options as the paragraph block rather than trying to automatically set a contrasting text colour, ie. provide both a text and background colour setting option, and warn the user if a low accessibility colour combination is chosen.

**Comment by glendaviesnz** [comment]
Pushed a revision which just adds a standard text colour selector, and refactors to use useColors instead of withColors. Would be good to get some feedback on:

- Whether it is seen as a good idea to be able to set the text colour at the group block level
- Whether it is better to just let the user manually do this, or to try and add some auto setting of a contrasting colour if a custom background colour us set

**Comment by karmatosed** [comment]
If I am right the design feedback here would be on whether to expect it to work without someone doing anything? For me, the answer is yes. I think it's fine as you an always go in and adjust yourself. I hope I am understanding this correctly though as to what the exact feedback is needed, if not please let me know.

**Comment by shaunandrews** [comment]
Text color for the group block makes sense to me.

> Whether it is better to just let the user manually do this, or to try and add some auto setting of a contrasting colour if a custom background colour us set

Could we do both? Set a light text color if a dark background is select, but also let users choose a custom color - showing the contrast alert if needed.

**Comment by richtabor** [comment]
Just to clarify behavior: If you go add a color to the Paragraph block within a Group block, does that applied color override the Group block's applied `TextColor`? Or does the Group block's `TextColor` overrule any subsequently colors applied to text within it?

We've experimented with both applications on CoBlocks, but I lean towards letting subsequent blocks handle their color applications (or better yet, pass down color selections from parent blocks). This way, if you change the nested Paragraph block color, it affects the block itself. 

Here's why:

If we went with a behavior where the Group block's `TextColor` wins every time, you end up with a UX scenario where you do not see the actual text color applied to the paragraph, even though the controls exist. 

If we went with a behavior where the Group block's `TextColor` is secondary to applied colors within nested blocks, you could end up in a UX scenario where if you've applied colors to some nested blocks, but are left not knowing 100% if changing the `TextColor` on the Group block will actually change the color of the nested blocks. 

Either way, it's not a fantastic experience. You actually end up having to mentally double-check that what you're doing will actually accomplish what you're wanting.

**Comment by ZebulanStanphill** [comment]
Logically, I think it makes sense for nested block controls to take precedence over the parent. Similar to how applying CSS styles work, a text color applied to a Paragraph should have a higher priority than the text color applied to its parent block.

I do think text controls on the Group block would be a good idea. It's not DRY (don't repeat yourself) to have to change all the Paragraphs in a section individually. Additionally, there's no guarantee that the nested blocks even have text color options (e.g. the List block currently lacks text color options). Actually, I think a Group block text color control is far more useful than a Paragraph text control.

Additionally, I think that since one of the primary uses of the Group block is to apply regional colors, we could surface both the background and text color options in the toolbar rather than the inspector. Remember that important block controls should not be hidden in the inspector.

**Comment by shaunandrews** [comment]
> Similar to how applying CSS styles

Exactly this — it's the cascading bit of CSS. I see it making logical sense, and it maps to an existing pattern outside of Gutenberg (but very fundamental to the web.) Document editors like MS Word also do similar things.

**Comment by kwight** [comment]
This works great @glendaviesnz ; I'd agree with others that adding the text controls and not letting the group block colors to clobber child block settings is the sensical way to go.

It _is_ strange to click into a child block and its colors settings, and to see no active values when there are obviously active values from the group block. I wonder if even a notice similar to the contrast notice could explain to the user "These colors are currently determined by this block's Group Block. Setting colors here will override the Group Block's colors for this block." Uh, or something.

**Comment by shaunandrews** [comment]
Some sort of notice _could_ be helpful. Though, I wonder if we'd have to show something for both background and text color on every child block that inherits colors from a parent? If so, seems like it could get a little too heavy.

image

**Comment by mapk** [comment]
It makes sense to me for the nested block's color selection to override the Group block's color selection. Maybe we can reduce some of the messaging by the scenario below:

**Scenario:**
I set the Group block's text color to be red. 

If a nested Paragraph block does NOT have text color specified, then the text color turns red changing the nested Paragraph block's text color value to red. This way there's no messaging needed, I just see red as the selected color for my Paragraph block.

If a Paragraph block does have text color specified (green), then it remains green, overriding the Group block's text color. If I were to "clear" this Paragraph's text color, it would automatically inherit the Group block's text color setting.

So in this instance, the Group block's text color just passes that selection down to nested blocks and is reflected there as well.

**Comment by ZebulanStanphill** [comment]
@mapk How would that work for multiple levels of nesting? Also, how would that work with blocks that don't already have their own text color option (e.g. List, Latest Posts, etc.)? Also, moving the blocks out of the Group would cause their text color to remain the same as when they were in the Group, which seems wrong to me.

**Comment by gwwar** [comment]
In terms of implementation will we be going with inline styles set on the elements? If so, this should handle cascading properly with arbitrarily nested elements. (Where if a child has a property set, this takes precedence over the parent.

The downside to this though, is that anything else attempting to override styling here would have trouble with the high CSS specificity (no number of classes or ids would be able to override).

**Comment by glendaviesnz** [comment]
Thanks for the great feedback everyone. 

The second iteration of this PR works the way most are suggesting it should, ie.

- Text colour selected on group the group block applies to all child blocks, unless the child block has specified its own colour, in which case this overrides the parent colour 
- This is handled with default css cascading, so no style attributes are actually added to the child blocks - so if the child block is move out of the current parent block the text colour will no longer be applied

Because of the fact that no attributes are applied to the children it will be difficult to show an indication that parent text colour styles are being applied without pushing down props through multiple levels of nested children, or some sore of redux plumbing. 

It seems like a shame to lose the simplicity of css here, which does the bulk of what we want by default, by adding extra plumbing to show a parents style settings in a child components colour selector. 

I am going to suggest that this PR is tidied up to just add the text colour option to the group, so at least people can set a suitable text colour for any background they select without having to manually set it on all the children. Then we could iterate back over this in further PRs to see if this can be easily indicated/messaged in child components, and if suitably contrasting text colours can be auto selected on custom background colour change.

**Comment by glendaviesnz** [comment]
Updated the description to better match the latest iteration and to include some of the feedback to-date.

**Comment by glendaviesnz** [comment]
> > Whether it is better to just let the user manually do this, or to try and add some auto setting of a contrasting colour if a custom background colour us set
> 
> Could we do both? Set a light text color if a dark background is select, but also let users choose a custom color - showing the contrast alert if needed.

Both would certainly be possible, but on looking at it, the new useColors hook seems like a better place to implement an automatic contrasting text selection option, rather than in the groups component specifically, that way it would be available to any component with useColours - have add a note about this to the PR description, in case someone wants to file a follow up issue about this.

**Comment by mapk** [comment]
> How would that work for multiple levels of nesting? 

It works just like I mentioned. The Group passes down the color if one isn't already set. I imagine if a user is setting the color on a Group block, their hoping to effect all text colors within it... unless some are already set. But if they wanted to effect those too, maybe there's a solution in the UI that allows this.

> Also, how would that work with blocks that don't already have their own text color option (e.g. List, Latest Posts, etc.)? 

Good question! In that case it would be helpful to show the Group block's text color option as this PR indicates.

> Also, moving the blocks out of the Group would cause their text color to remain the same as when they were in the Group, which seems wrong to me.

It doesn't bother me to move a block that I purposefully changed text color on out of a Group, and have it keep the same color I already purposefully set on it. That feels alright to me.

**Comment by glendaviesnz** [comment]
@youknowriad, @talldan, @Soean, @ajitbohra, @jorgefilipecosta  - before I rebase this PR, and maybe add some tests, are you able to confirm if this is feature that you would want added into core? Thanks.

**Comment by youknowriad** [comment]
I feel like I'm missing context here. This has been discussed I believe when the block was first added. Maybe @getdave @talldan @mtias @kjellr and others could tell us why it wasn't added initially, if it was just left as a follow-up or if there was a strong argument against it.

**Comment by jasmussen** [comment]
For what it's worth, I'm personally more of a fan of adding color controls to containers like this, rather than duplicating the same UI for each individual UI.

**Comment by glendaviesnz** [comment]
I just rebased this, and updated the colorDetector to match the latest version. I have done some testing and it all seems to work as it should. If someone else had time to do some testing and review the code that would be great, thanks 😃

**Comment by talldan** [comment]
I don't recall any specific conversation about this on the original PR. 

I had a quick look, but the PR is pretty huge (https://github.com/WordPress/gutenberg/pull/13964).

The description mentions it was avoided to keep the original PR small, so probably just that.

**Comment by mapk** [comment]
I'm really stoked with how this is turning out! 👍 Here's a gif of how it looks/works.

groupcolors 2020-01-16 18_33_25

**Comment by glendaviesnz** [comment]
> I'm really stoked with how this is turning out

Thanks for the detailed testing! People complain about CSS, but this is a scenario where the 'cascade' is out friend :-)

**Review by youknowriad — APPROVED** [review]
This is looking great for me. Thanks for the addition. I also love the refactor with useColors.

File: packages/block-library/src/group/edit.js [hunk 1]
```diff
@@ -1,62 +1,52 @@
...
-			<div className={ classes } style={ styles }>
-				<div className="wp-block-group__inner-container">
-					<InnerBlocks
-						renderAppender={ ! hasInnerBlocks && InnerBlocks.ButtonBlockAppender }
-					/>
-				</div>
-			</div>
+			{ InspectorControlsColorPanel }
+			<BackgroundColor>
+				<TextColor>
```
Comment (youknowriad) [inline packages/block-library/src/group/edit.js]: It still seems that a single `ApplyColors` would be better as it's used that way in 90% of the use-cases. cc @epiqueras @jorgefilipecosta

Comment (epiqueras) [reply packages/block-library/src/group/edit.js]: It could be an additional component that's returned from the hook.

**Comment by richtabor** [comment]
> 👍 Here's a gif of how it looks/works.

Looking good. One question though @mapk: With the gif as a reference, do you think it's confusing that the paragraph text within the group is "green", yet when you open its the `Colors` panel, the color is not reflected there? 

Also, we need to verify these cascading styles are properly reflected on the frontend (using both color classes and custom inline styles).

**Comment by mapk** [comment]
> With the gif as a reference, do you think it's confusing that the paragraph text within the group is "green", yet when you open its the Colors panel, the color is not reflected there?

Good question, @richtabor! The interaction here gets tricky. So currently when observing the color settings of the Paragraph block nested inside the Group block, we don't see the green color selected even though the Paragraph block has inherited green from the Group block color setting. 

I'm okay with this interaction right now because I think it gets more complex if we choose to show the green selected color setting being pushed to the nested blocks too. Like if I change the Paragraph block to red, and then click to "clear" that color setting, does it default to green again? That doesn't feel like a proper "clear" to me. I'm willing to be persuaded otherwise.

**Comment by mtias** [comment]
The interactions between these levels of color definition is precisely what the work of @ItsJonQ in #19611 aims to address. The "current color" on a paragraph being empty but still reflecting the color of a parent in the UI should be communicated somehow.

**Comment by glendaviesnz** [comment]
> The "current color" on a paragraph being empty but still reflecting the color of a parent in the UI should be communicated somehow.

Any thoughts on whether we should merge it as is and circle back on this, or try and resolve this first?

**Comment by ZebulanStanphill** [comment]
I think this is fine to merge as-is.

**Comment by glendaviesnz** [comment]
Doh - apologies for the poor commit message, was focused on the more detailed message and overlooked that the first part was just the original branch name - I won't try a forced push to master just to try and amend this 😄

**Pull request merged**